### PR TITLE
Handle getting state abbreviation for possible secondary field

### LINKED ISSUE #5291: Address Standarization does not correct for State
**Describe the bug**

Addresses that are entered with the wrong state are not corrected during the address standardization window. 

1. Enter a correct Address in WA. Enter any other state than WA. 
2. When the standardization page asks you to select a correct address, you can select an address with the original incorrect state. 

**Expected behavior**

All Seattle addresses should at least correct to WA for the state. 

**Screenshots**

**Desktop (please complete the following information):**

**Smartphone (please complete the following information):**

**Additional context**

### PR DESCRIPTION
**Pull request by gwendolyngoetz**
### Description

Get the state abbreviation from the attributes json node. On public ESRI this is under the RegionAbbr field, but some custom implementations may have it under the Region field.

## Release notes

The title of the pull request will be used as the default release notes description. If more detail is needed to communicate to partners the scope of the PR's changes, use this release notes section.

#### General

Read the full guidelines for PRs [here](https://docs.civiform.us/contributor-guide/developer-guide/technical-contribution-guide#guidelines-for-pull-requests)

- [x] Added the correct label: < feature | bug | dependencies | infrastructure | ignore-for-release | database >
- [x] Created unit and/or browser tests which fail without the change (if possible)
- [x] Performed manual testing (Chrome and Firefox if it includes front-end changes)

### Issue(s) this completes

Fixes  #5291

### CODE REVIEW

File: server/app/services/geo/esri/EsriClient.java [hunk 1]
```diff
@@ -179,6 +177,35 @@ public CompletionStage<AddressSuggestionGroup> getAddressSuggestions(Address add
...
   }
 
+  /**
+   * Get the state abbreviation from the attributes json node. on Public ESRI this is under the
+   * RegionAbbr field, but some custom implementations may have it under the Region field.
+   */
+  private String getStateAbbreviationFromAttributes(JsonNode attributes, Address address) {
+    BiFunction<JsonNode, String, String> getValue =
+        (jsonNode, keyName) ->
+            (jsonNode.get(keyName) == null || jsonNode.get(keyName).isNull())
```
Comment (dkatzz) [inline server/app/services/geo/esri/EsriClient.java:187]: **question:** Why do we need both of these null checks? Should it be isEmpty like above?

File: server/app/services/geo/esri/EsriClient.java [hunk 2]
```diff
@@ -179,6 +177,35 @@ public CompletionStage<AddressSuggestionGroup> getAddressSuggestions(Address add
...
+                ? ""
+                : jsonNode.get(keyName).asText();
+
+    // Start with getting the value from the RegionAbbr, this is the
+    // default used by ESRI for state abbreviations
+    String result = getValue.apply(attributes, "RegionAbbr");
+
+    // Assume if it's not two characters it is either empty or the
+    // full state name and pull from the Region value which
```
Comment (dkatzz) [inline server/app/services/geo/esri/EsriClient.java:196]: **suggestion:** Remove which (or finish sentence)

**Review by dkatzz — APPROVED** [review]

**Comment by dkatzz** [comment]
Nit: you can remove some of the sections in the PR description that aren't being used.

File: server/app/services/geo/esri/EsriClient.java (hunk 1 above)
Comment (gwendolyngoetz) [reply server/app/services/geo/esri/EsriClient.java:187]: One checks if the object is null, the other checks for a null node type. I'm going to write some more tests, I think using isEmpty is actually a bug.

File: server/app/services/geo/esri/EsriClient.java (hunk 2 above)
Comment (gwendolyngoetz) [reply server/app/services/geo/esri/EsriClient.java]: done

File: server/app/services/geo/esri/EsriClient.java (hunk 1 above)
Comment (gwendolyngoetz) [reply server/app/services/geo/esri/EsriClient.java]: @dkatzz, Ok fixed a whole different bug than I planned. Turns out `isEmpty` on scalars always returns true. It's really looking at children in arrays/objects. The `isNull` method is the correct thing to use. I've pull things apart a little more to fix it and switched to a more targeted series of tests to check the mapping.

File: server/app/services/geo/esri/EsriClient.java [hunk 3]
```diff
@@ -179,6 +165,75 @@ public CompletionStage<AddressSuggestionGroup> getAddressSuggestions(Address add
             });
   }
 
+  /** Maps the json values within the node to an address */
```
Comment (dkatzz) [inline server/app/services/geo/esri/EsriClient.java:168]: **nitpick:** Capitalize to JSON (here and below)

File: server/app/services/geo/esri/EsriClient.java [hunk 4]
```diff
@@ -179,6 +165,75 @@ public CompletionStage<AddressSuggestionGroup> getAddressSuggestions(Address add
...
+        // returned by the ESRI instance and we don't want to lose the value.
+        .setLine2(getNodeTextValueOrDefault(attributes, "SubAddr", address::getLine2))
+        .setCity(getNodeTextValueOrEmpty(attributes, "City"))
+        .setState(getStateAbbreviationFromAttributes(attributes, address))
+        .setZip(getNodeTextValueOrEmpty(attributes, "Postal"))
+        .build();
+  }
+
+  /**
+   * Get the state abbreviation from the attributes json node. on Public ESRI this is under the
```
Comment (dkatzz) [inline server/app/services/geo/esri/EsriClient.java:185]: **nitpick:** Capitalize to On

File: server/test/services/geo/esri/EsriClientTest.java [hunk 5]
```diff
@@ -177,4 +182,83 @@ public void getAddressSuggestionsWithError() throws Exception {
     assertThat(group.getWellKnownId()).isEqualTo(0);
     assertThat(group.getOriginalAddress()).isEqualTo(address);
   }
+
```
Comment (dkatzz) [inline server/test/services/geo/esri/EsriClientTest.java:185]: **suggestion:** Could you add a test where both Region and RegionAbbr are both > 2 letters?

File: server/test/services/geo/esri/EsriClientTest.java [hunk 6]
```diff
@@ -177,4 +182,83 @@ public void getAddressSuggestionsWithError() throws Exception {
...
+            + "  \"Region\": \"WA\","
+            + "  \"RegionAbbr\": \"Washington\","
+            + "  \"Postal\": \"11111-expected\""
+            + "}}";
+
+    runMapAddressAttributesJson(
+        json, "street-expected", "line2-user", "city-expected", "WA", "11111-expected");
+  }
+
+  private void runMapAddressAttributesJson(
```
Comment (dkatzz) [inline server/test/services/geo/esri/EsriClientTest.java:237]: **suggestion:** I'm not sure what the best name is, but thoughts on calling this something like runAndVerifyMapAddressAttributesJson or something? I think with the name its unclear this function is doing the asserts as well.

File: server/app/services/geo/esri/EsriClient.java (hunk 1 above)
Comment (dkatzz) [reply server/app/services/geo/esri/EsriClient.java]: Ah makes sense - thanks for the update! The way you pulled things apart LGTM - just some minor comments.

**Pull request merged**